scalanet/src/io/iohk/scalanet/peergroup/kademlia/KBuckets.scala

```diff
@@ -1,46 +1,66 @@
 package io.iohk.scalanet.peergroup.kademlia
 
-import java.util.concurrent.ConcurrentSkipListSet
+import java.time.Clock
 
 import scodec.bits.BitVector
 
-import scala.collection.JavaConverters._
-
 /**
   * Skeletal kbucket implementation.
+  *
   * @param baseId the nodes own id.
   */
-class KBuckets(val baseId: BitVector) {
-
-  private val nodeIds =
-    new ConcurrentSkipListSet[BitVector](new XorOrdering(baseId)).asScala
+class KBuckets private (val baseId: BitVector, val clock: Clock, val buckets: List[TimeSet[BitVector]]) {
 
-  add(baseId)
+  def this(baseId: BitVector, clock: Clock) =
+    this(baseId, clock, List.fill(baseId.length.toInt)(TimeSet[BitVector](clock)))
 
   /**
     * Find the n nodes closest to nodeId in kBuckets.
     * Return the resulting node records sorted by distance from the nodeId.
     * The indices into the kBuckets are defined by their distance from referenceNodeId.
     */
   def closestNodes(nodeId: BitVector, n: Int): List[BitVector] = {
-    // replace with routing tree described in kademlia paper...
-    nodeIds.toList.sorted(new XorOrdering(nodeId)).take(n)
+    val ordering = new XorOrdering(nodeId)
```

Comment:
This implementation is not correct:
```
def bv(s: String) = BitVector.fromBin(s).get
val kbuckets = new KBuckets(bv("00000"), Clock.systemUTC).add(bv("00001")).add(bv("00010")).add(bv("00100")).add(bv("01000")).add(bv("10000"))
kbuckets.closestNodes(bv("00111"), 3).map(_.toBin)
```
Returns `List(00100, 00010, 00000)`, should be `List(00100, 00010, 00001)`.

Your bucket ordering is incorrect, you first have to take the bucket `nodeId` falls into, then look into lower subbuckets and only then check higher ones.

Here is the pseudocode of correct ordering (of buckets; non-tailrec):
```
def sortBuckets(nodeId: BitVector) = {
  // loop(pos) returns sorted list of buckets with id >= p
  // ordering has following property: if one bucket appears before the next, all of nodes inside the former are closer to nodeid than ones in the latter
  def loop(pos: Int) = {
    if (p == buckets.size) {
      Nil
    } else {
      if (nodeId(pos) == baseId(pos)) {
        // buckets returned by loop(p + 1) share (p + 1)-bit prefix with baseId, so their nodes xorred with nodeId share (p + 1)-bit prefix with (nodeId xor baseId)  - with the (p + 1)-th byte being 0
        // bucket pos share p-bit prefix and differs on (p + 1) bit, so its elements xorred with nodeId give higher metrics value than ones from loop(p + 1), as they share p-bit prefixes and differ on (p + 1)-th bit, which is 0 for elements from loop(pos + 1) and 1 from p-th bucket
        loop(pos + 1) ++ List(buckets(p))
      } else {
        // elements from buckets(p) and from loop(p + 1) buckets xorred with nodeId share p-bit prefix and differ on (p + 1)-th bit, with elements from buckets(p) giving 0 (as that is the first place where they mismatch with baseId and baseId(p) != nodeId(p)) and ones from buckets returned by loop(p + 1) having 1 - so the former are closer in regards to XOR metric
        buckets(p) :: loop(pos + 1)
      }
    }
  }
}
```

We can have a chat on that tomorrow or I can write the code or you can use unoptimized implementation and leave optimization for next tickets - all options are fine by me.